engine/utils/collections/HeapArray.h

```diff
@@ -0,0 +1,484 @@
+        ArrayUtils::CopyData(data, rawPtr, sizeof(T) * size);
+
+        count = size;
+    }
+
+    /**
+     * @brief A copy constructor for the HeapArray
+     * @param other the array to be copied
+     */
+    MHArray(const MHArray<T>& other) : size {other.size}, count {other.count}
+    {
+        size = other.size;
+
+        if (other.data == nullptr && other.size == 0) return;
+
+        auto stateMaskCount = BitUtils::CalculateBitSetSize(other.size);
+
+        AllocateMemory(size, stateMaskCount);
+
+        ArrayUtils::CopyData(data, other.data, sizeof(T) * size);
+
+        bitField = other.bitField;
+
+        count = other.count;
+    }
+
+    /**
+     * @brief A move constructor for the HeapArray
+     * @param other the array to be moved
+     */
+    MHArray(MHArray<T>&& other) noexcept
+    {
+        Swap(std::move(other));
+    }
+
+    /**
+     * @brief The destructor for the HeapArray
+     */
+    ~MHArray()
+    {
+        Destroy();
+    }
+
+    /**
+     * @brief Deletes any memory that's been allocated and resets the array to default values
+     */
+    void Destroy()
+    {
+        DeallocateMemory();
+        ResetValues();
+    }
+
+    /**
+     * @brief An overload for the subscript operator. Gets the element at the specified index
+     * @param index the index to search for
+     * @return a const reference to the element at the given index
+     */
+    const T& operator[](const size_t& index) const
+    {
+        return Get(index);
+    }
+
+    /**
+     * @brief An overload for the subscript operator. Gets the element at the specified index.
+     * If the index has not previously been assigned and we use the subscript to assign a value
+     * (i.e: array[0] = x), then we need to increment the count of items in the array
+     * @param index the index to search for
+     * @return a reference to the element at the given index
+     */
+    T& operator[](const size_t& index)
+    {
+        assert(data && "Trying to index into MHArray when the array has not been initialised!");
+        count += !bitField.IsSet(index + 1);
+        bitField.SetBit(index + 1);
+
+        return Get(index);
+    }
+
+    /**
+     * @brief Copy assignment operator. Copies the content of `other` into the current array
+     * @param other the array to be copied
+     * @return the array with the copied elements
+     */
+    MHArray<T>& operator=(const MHArray<T>& other)
+    {
+        // Don't copy if the objects are the same
+        if (this == &other) return *this;
+
+        Destroy();
+
+        Copy(other);
+
+        return *this;
+    }
+
+    /**
+     * @brief Move assigment operator. Moves the contents of `other` into the current array
+     * @param other the array to be moved
+     * @return the new array with the moved contents
+     */
+    MHArray<T>& operator=(MHArray<T>&& other) noexcept
+    {
+        if (this == &other) return *this;
+
+        // Deallocate all existing memory
+        Destroy();
```

Comment:
The swap will cause a destroy implicitly. I don't think this is needed

Reply:
Good catch. Removed.